### ISSUE TITLE
Creation of existing container leads to abandoned CLOSE_WAIT socket

### ISSUE THREAD

**Issue by awskii** [gophercloud/gophercloud#1015]
Hi.

When I send request to Create (put) container, which is already exists, i've got abandoned socket in close_wait state, which means, you don't send SYN to peer.

Proof:
```$ lsof -a -p 11520
11520 awskii    4u  a_inode   0,13        0     11625 [eventpoll]
11520 awskii    5u     IPv4 157447      0t0       TCP localhost:http-alt (LISTEN)
11520 awskii    6u     IPv6 159056      0t0       TCP *:iprop (LISTEN)
```
[create container with existing name via `containers.Create` method]
```
11520 awskii    4u  a_inode   0,13        0     11625 [eventpoll]
11520 awskii    5u     IPv4 157447      0t0       TCP localhost:http-alt (LISTEN)
11520 awskii    6u     IPv6 159056      0t0       TCP *:iprop (LISTEN)
11520 awskii    7u     IPv6 160355      0t0       TCP localhost:iprop->localhost:54010 (ESTABLISHED)
11520 awskii    8u     IPv4 157521      0t0       TCP W:42054->92.53.68.203:https (CLOSE_WAIT)
```
[yet another existing container creation]
```
11520 awskii    4u  a_inode   0,13        0     11625 [eventpoll]
11520 awskii    5u     IPv4 157447      0t0       TCP localhost:http-alt (LISTEN)
11520 awskii    6u     IPv6 159056      0t0       TCP *:iprop (LISTEN)
11520 awskii    7u     IPv6 160355      0t0       TCP localhost:iprop->localhost:54010 (ESTABLISHED)
11520 awskii    8u     IPv4 157521      0t0       TCP W:42054->92.53.68.203:https (CLOSE_WAIT)
11520 awskii    9u     IPv4 159367      0t0       TCP W:42108->92.53.68.203:https (CLOSE_WAIT)
```

Sockets are abandoned, and not closing for days.

Using `resp.Body.Close()` in `containers.Create` method solving this issue.

**Comment by awskii** [comment]
In objects.Create you are using Close function. so, i think, it's a bug that you don't use it with containers creation.
This issue was there for months if no more and it was really hard to detect such careless socket usage.

**Comment by jtopjian** [comment]
@awskii Thanks for reporting this.

Can you provide some further details about how to reproduce the problem? I'm unable to reproduce this.

I'm using the following test to trigger `containers.Create`:

```go
func TestDupCreate(t *testing.T) {                 
  client, err := clients.NewObjectStorageV1Client()
  if err != nil {                                  
    t.Fatalf("Unable to create client: %v", err)   
  }                                                
                                                   
  res := containers.Create(client, "foo", nil)     
  th.AssertNoErr(t, res.Err)                       
}                                                  
```

And then I do the following:

```shell
$ go test -v -tags "fixtures acceptance" -run "TestDupCreate" github.com/gophercloud/gophercloud/acceptance/openstack/objectstorage/v1

$ netstat -nap  2>&1 > ~/foo-1.txt

$ go test -v -tags "fixtures acceptance" -run "TestDupCreate" github.com/gophercloud/gophercloud/acceptance/openstack/objectstorage/v1

$ netstat -nap  2>&1 > ~/foo-2.txt

$ diff ~/foo-1.txt ~/foo-2.txt
5a6                                                                                
> tcp        0      0 192.168.244.102:57526   192.168.244.216:5000    TIME_WAIT   -
```

**Comment by awskii** [comment]
@jtopjian, that's code snippet to reproduce issue. Fill `login` and `password` variables and change `options. IdentityEndpoint` by yours and check it out.

```go
package main

import (
	"log"
	"net/url"
	"time"

	"github.com/gophercloud/gophercloud"
	"github.com/gophercloud/gophercloud/openstack"
	"github.com/gophercloud/gophercloud/openstack/objectstorage/v1/containers"
)

func main() {
	login, password := "", ""

	options := gophercloud.AuthOptions{
		IdentityEndpoint: "https://auth.selcdn.ru/v3",
		TenantID:         login,
		UserID:           login,
		Password:         password,
		AllowReauth:      true, // session caching
	}

	endpointOpts := gophercloud.EndpointOpts{
		Name: "swift",
		Type: "object-store",
	}

	provider, err := openstack.NewClient(options.IdentityEndpoint)
	if err != nil {
		log.Println(err)
	}

	if err := openstack.Authenticate(provider, options); err != nil {
		log.Println(err)
	}

	swift, err := openstack.NewObjectStorageV1(provider, endpointOpts)
	if err != nil {
		log.Println(err)
	}
	log.Println("client obtained")

	containerName := "cw"

	res := containers.Create(swift, url.PathEscape(containerName), &containers.CreateOpts{})
	if err := res.Err; err != nil {
		log.Println(err)
	}
	res = containers.Create(swift, url.PathEscape(containerName), &containers.CreateOpts{})
	if err := res.Err; err != nil {
		log.Println(err)
	}
	log.Println("ready")
	time.Sleep(time.Minute)
}
```

```
$ lsof -a -p $(pidof cw)

COMMAND   PID   USER   FD     TYPE             DEVICE   SIZE/OFF       NODE NAME
cw      18178 awskii  cwd      DIR                1,5        192 8593286906 /Users/awskii/cw
...
cw      18178 awskii    5u    IPv4 0x70b88b75c34a3161        0t0        TCP 192.168.0.101:52913->92.53.68.205:https (CLOSE_WAIT)

**Comment by awskii** [comment]
Bug reproducing both mac\linux.

**Comment by jtopjian** [comment]
@awskii confirmed! :)

Curiously, I had to compile into a binary before I was able to reproduce. But you're right.

The reason is because the Body of the result when creating a container which already exists is not empty. The body contains:

```
<html><h1>Accepted</h1><p>The request is accepted for processing.</p></html>
```

And because of this, the body is left open expecting something to read it.

I think the easiest solution here is:

```diff
--- a/openstack/objectstorage/v1/containers/requests.go
+++ b/openstack/objectstorage/v1/containers/requests.go
@@ -107,6 +107,7 @@ func Create(c *gophercloud.ServiceClient, containerName string, opts CreateOptsB
        })
        if resp != nil {
                r.Header = resp.Header
+               resp.Body.Close()
        }
        r.Err = err
        return
```

But I'm wondering if there's a better place to do this. @jrperritt thoughts?

**Comment by awskii** [comment]
@jtopjian, as I said above: 
> Using resp.Body.Close() in containers.Create method solving this issue.
I think, better place for that can be in Request method, just to bring all work with http request together. But we can't rely only on Content-Length header to decide if we need to close body. Also, in `object.Create` response body is closed right there.

**Comment by jtopjian** [comment]
> as I said above:

Yes, I was just showing an exact area in `containers.Create` where the fix could go 😄 

> I think, better place for that can be in Request method, just to bring all work with http request together. 

Yeah, I thought about this, too, but I'm not sure if this would cause any unwanted side-effects. At the moment, this seems to only be an issue with containers, so I'm leaning on fixing this locally with containers.

**Comment by jrperritt** [comment]
> so I'm leaning on fixing this locally with containers

I agree

**Comment by jtopjian** [comment]
This has been fixed in #1016. @awskii thank you very much for reporting this problem :)
